src/muport.js

```diff
@@ -112,21 +112,27 @@ class MuPort {
     if (delegateDids.length !== 3) throw new Error('Must provide exactly 3 DIDs')
     // generate new recoveryNetwork
     const didsPublicKeys = await Promise.all(delegateDids.map(async did => (await MuPort.resolveIdentityDocument(did)).asymEncryptionKey))
-    this.document.recoveryNetwork = await this.keyring.createShares(delegateDids, didsPublicKeys)
+    // seems like the best way to do a deep clone
+    const newDocument = JSON.parse(JSON.stringify(this.document))
+    newDocument.recoveryNetwork = await this.keyring.createShares(delegateDids, didsPublicKeys)
     // save guardians
-    this.document.symEncryptedData.symEncDids = delegateDids.map((did) => this.keyring.symEncrypt(didToBuffer(did)))
-    this.documentHash = await ipfs.addJSONAsync(this.document)
+    newDocument.symEncryptedData = newDocument.symEncryptedData || {}
+    newDocument.symEncryptedData.symEncDids = delegateDids.map((did) => this.keyring.symEncrypt(didToBuffer(did)))
+    let newDocumentHash = await ipfs.addJSONAsync(newDocument)
     // prepare ethereum tx
     const address = this.keyring.getManagementAddress()
-    const txParams = await this.ethUtils.createPublishTxParams(this.documentHash, address)
+    const txParams = await this.ethUtils.createPublishTxParams(newDocumentHash, address)
     const costInEther = this.ethUtils.calculateTxCost(txParams)
     const signedTx = this.keyring.signManagementTx(txParams)
 
     return {
       address,
       costInEther,
       finishUpdate: async () => {
-        await this.ethUtils.sendRawTx(signedTx)
+        const txHash = await this.ethUtils.sendRawTx(signedTx)
+        await this.ethUtils.waitForTx(txHash)
+        this.document = newDocument
```

Comment:
Ok so now you can cancel at any point before calling `finishUpdate`, which is good. What happens if the above transaction fails? Should we ideally be checking the transaction receipt and only updating the document/hash if the transaction is successful?

Reply:
Good point. I guess we can just put the `waitForTx` call and everything after it in a `try` clause?

Reply:
Hmm not sure, I think we can start by just worrying about those.